// @flowgram.ai/runtime-interface/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  FlowGramAPIMethod: () => FlowGramAPIMethod,
  FlowGramAPIModule: () => FlowGramAPIModule,
  FlowGramAPIName: () => FlowGramAPIName,
  FlowGramAPINames: () => FlowGramAPINames,
  FlowGramAPIs: () => FlowGramAPIs,
  FlowGramNode: () => FlowGramNode,
  IEngine: () => IEngine,
  IExecutor: () => IExecutor,
  IValidation: () => IValidation,
  ServerInfoDefine: () => ServerInfoDefine,
  TaskCancelDefine: () => TaskCancelDefine,
  TaskReportDefine: () => TaskReportDefine,
  TaskResultDefine: () => TaskResultDefine,
  TaskRunDefine: () => TaskRunDefine,
  ValidationDefine: () => ValidationDefine,
  WorkflowPortType: () => WorkflowPortType,
  WorkflowStatus: () => WorkflowStatus,
  WorkflowVariableType: () => WorkflowVariableType
});
module.exports = __toCommonJS(src_exports);

// src/api/validation/index.ts
var import_zod = __toESM(require("zod"));

// src/api/constant.ts
var FlowGramAPIMethod = /* @__PURE__ */ ((FlowGramAPIMethod2) => {
  FlowGramAPIMethod2["GET"] = "GET";
  FlowGramAPIMethod2["POST"] = "POST";
  FlowGramAPIMethod2["PUT"] = "PUT";
  FlowGramAPIMethod2["DELETE"] = "DELETE";
  FlowGramAPIMethod2["PATCH"] = "PATCH";
  return FlowGramAPIMethod2;
})(FlowGramAPIMethod || {});
var FlowGramAPIName = /* @__PURE__ */ ((FlowGramAPIName2) => {
  FlowGramAPIName2["ServerInfo"] = "ServerInfo";
  FlowGramAPIName2["TaskRun"] = "TaskRun";
  FlowGramAPIName2["TaskReport"] = "TaskReport";
  FlowGramAPIName2["TaskResult"] = "TaskResult";
  FlowGramAPIName2["TaskCancel"] = "TaskCancel";
  FlowGramAPIName2["Validation"] = "Validation";
  return FlowGramAPIName2;
})(FlowGramAPIName || {});
var FlowGramAPIModule = /* @__PURE__ */ ((FlowGramAPIModule2) => {
  FlowGramAPIModule2["Info"] = "Info";
  FlowGramAPIModule2["Task"] = "Task";
  FlowGramAPIModule2["Validation"] = "Validation";
  return FlowGramAPIModule2;
})(FlowGramAPIModule || {});

// src/api/validation/index.ts
var ValidationDefine = {
  name: "Validation" /* Validation */,
  method: "POST" /* POST */,
  path: "/validation",
  module: "Validation" /* Validation */,
  schema: {
    input: import_zod.default.object({
      schema: import_zod.default.string()
    }),
    output: import_zod.default.object({
      valid: import_zod.default.boolean(),
      nodeErrors: import_zod.default.array(
        import_zod.default.object({
          message: import_zod.default.string(),
          nodeID: import_zod.default.string()
        })
      ),
      edgeErrors: import_zod.default.array(
        import_zod.default.object({
          message: import_zod.default.string(),
          edge: import_zod.default.object({
            sourceNodeID: import_zod.default.string(),
            targetNodeID: import_zod.default.string(),
            sourcePortID: import_zod.default.string().optional(),
            targetPortID: import_zod.default.string().optional()
          })
        })
      )
    })
  }
};

// src/api/task-run/index.ts
var import_zod3 = __toESM(require("zod"));

// src/api/schema.ts
var import_zod2 = __toESM(require("zod"));
var WorkflowIOZodSchema = import_zod2.default.record(import_zod2.default.string(), import_zod2.default.any());
var WorkflowSnapshotZodSchema = import_zod2.default.object({
  id: import_zod2.default.string(),
  nodeID: import_zod2.default.string(),
  inputs: WorkflowIOZodSchema,
  outputs: WorkflowIOZodSchema.optional(),
  data: WorkflowIOZodSchema,
  branch: import_zod2.default.string().optional()
});
var WorkflowStatusZodShape = {
  status: import_zod2.default.string(),
  terminated: import_zod2.default.boolean(),
  startTime: import_zod2.default.number(),
  endTime: import_zod2.default.number().optional(),
  timeCost: import_zod2.default.number()
};
var WorkflowStatusZodSchema = import_zod2.default.object(WorkflowStatusZodShape);
var WorkflowZodSchema = {
  Inputs: WorkflowIOZodSchema,
  Outputs: WorkflowIOZodSchema,
  Status: WorkflowStatusZodSchema,
  Snapshot: WorkflowSnapshotZodSchema,
  NodeReport: import_zod2.default.object({
    id: import_zod2.default.string(),
    ...WorkflowStatusZodShape,
    snapshots: import_zod2.default.array(WorkflowSnapshotZodSchema)
  })
};

// src/api/task-run/index.ts
var TaskRunDefine = {
  name: "TaskRun" /* TaskRun */,
  method: "POST" /* POST */,
  path: "/task/run",
  module: "Task" /* Task */,
  schema: {
    input: import_zod3.default.object({
      schema: import_zod3.default.string(),
      inputs: WorkflowZodSchema.Inputs
    }),
    output: import_zod3.default.object({
      taskID: import_zod3.default.string()
    })
  }
};

// src/api/task-result/index.ts
var import_zod4 = __toESM(require("zod"));
var TaskResultDefine = {
  name: "TaskResult" /* TaskResult */,
  method: "GET" /* GET */,
  path: "/task/result",
  module: "Task" /* Task */,
  schema: {
    input: import_zod4.default.object({
      taskID: import_zod4.default.string()
    }),
    output: WorkflowZodSchema.Outputs
  }
};

// src/api/task-report/index.ts
var import_zod5 = __toESM(require("zod"));
var TaskReportDefine = {
  name: "TaskReport" /* TaskReport */,
  method: "GET" /* GET */,
  path: "/task/report",
  module: "Task" /* Task */,
  schema: {
    input: import_zod5.default.object({
      taskID: import_zod5.default.string()
    }),
    output: import_zod5.default.object({
      id: import_zod5.default.string(),
      inputs: WorkflowZodSchema.Inputs,
      outputs: WorkflowZodSchema.Outputs,
      workflowStatus: WorkflowZodSchema.Status,
      reports: import_zod5.default.record(import_zod5.default.string(), WorkflowZodSchema.NodeReport)
    })
  }
};

// src/api/task-cancel/index.ts
var import_zod6 = __toESM(require("zod"));
var TaskCancelDefine = {
  name: "TaskCancel" /* TaskCancel */,
  method: "PUT" /* PUT */,
  path: "/task/cancel",
  module: "Task" /* Task */,
  schema: {
    input: import_zod6.default.object({
      taskID: import_zod6.default.string()
    }),
    output: import_zod6.default.object({
      success: import_zod6.default.boolean()
    })
  }
};

// src/api/server-info/index.ts
var import_zod7 = __toESM(require("zod"));
var ServerInfoDefine = {
  name: "ServerInfo" /* ServerInfo */,
  method: "GET" /* GET */,
  path: "/info",
  module: "Info" /* Info */,
  schema: {
    input: import_zod7.default.undefined(),
    output: import_zod7.default.object({
      name: import_zod7.default.string(),
      runtime: import_zod7.default.string(),
      version: import_zod7.default.string(),
      time: import_zod7.default.string()
    })
  }
};

// src/api/define.ts
var FlowGramAPIs = {
  ["ServerInfo" /* ServerInfo */]: ServerInfoDefine,
  ["TaskRun" /* TaskRun */]: TaskRunDefine,
  ["TaskReport" /* TaskReport */]: TaskReportDefine,
  ["TaskResult" /* TaskResult */]: TaskResultDefine,
  ["TaskCancel" /* TaskCancel */]: TaskCancelDefine,
  ["Validation" /* Validation */]: ValidationDefine
};
var FlowGramAPINames = Object.keys(FlowGramAPIs);

// src/schema/constant.ts
var WorkflowPortType = /* @__PURE__ */ ((WorkflowPortType2) => {
  WorkflowPortType2["Input"] = "input";
  WorkflowPortType2["Output"] = "output";
  return WorkflowPortType2;
})(WorkflowPortType || {});
var WorkflowVariableType = /* @__PURE__ */ ((WorkflowVariableType2) => {
  WorkflowVariableType2["String"] = "string";
  WorkflowVariableType2["Integer"] = "integer";
  WorkflowVariableType2["Number"] = "number";
  WorkflowVariableType2["Boolean"] = "boolean";
  WorkflowVariableType2["Object"] = "object";
  WorkflowVariableType2["Array"] = "array";
  WorkflowVariableType2["Null"] = "null";
  return WorkflowVariableType2;
})(WorkflowVariableType || {});

// src/node/constant.ts
var FlowGramNode = /* @__PURE__ */ ((FlowGramNode2) => {
  FlowGramNode2["Root"] = "root";
  FlowGramNode2["Start"] = "start";
  FlowGramNode2["End"] = "end";
  FlowGramNode2["LLM"] = "llm";
  FlowGramNode2["code"] = "code";
  FlowGramNode2["Condition"] = "condition";
  FlowGramNode2["Loop"] = "loop";
  FlowGramNode2["Comment"] = "comment";
  FlowGramNode2["Group"] = "group";
  return FlowGramNode2;
})(FlowGramNode || {});

// src/runtime/engine/index.ts
var IEngine = Symbol.for("Engine");

// src/runtime/executor/executor.ts
var IExecutor = Symbol.for("Executor");

// src/runtime/status/index.ts
var WorkflowStatus = /* @__PURE__ */ ((WorkflowStatus2) => {
  WorkflowStatus2["Pending"] = "pending";
  WorkflowStatus2["Processing"] = "processing";
  WorkflowStatus2["Succeeded"] = "succeeded";
  WorkflowStatus2["Failed"] = "failed";
  WorkflowStatus2["Canceled"] = "canceled";
  return WorkflowStatus2;
})(WorkflowStatus || {});

// src/runtime/validation/index.ts
var IValidation = Symbol.for("Validation");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FlowGramAPIMethod,
  FlowGramAPIModule,
  FlowGramAPIName,
  FlowGramAPINames,
  FlowGramAPIs,
  FlowGramNode,
  IEngine,
  IExecutor,
  IValidation,
  ServerInfoDefine,
  TaskCancelDefine,
  TaskReportDefine,
  TaskResultDefine,
  TaskRunDefine,
  ValidationDefine,
  WorkflowPortType,
  WorkflowStatus,
  WorkflowVariableType
});
